packages/fxa-settings/src/pages/Signin/SigninRecoveryCode/container.tsx

```diff
@@ -0,0 +1,107 @@
+import { hardNavigateToContentServer } from 'fxa-react/lib/utils';
+import { MozServices } from '../../../lib/types';
+import SigninRecoveryCode from '.';
+import { Integration, useAuthClient } from '../../../models';
+import LoadingSpinner from 'fxa-react/components/LoadingSpinner';
+import { useMutation } from '@apollo/client';
+import { CONSUME_RECOVERY_CODE_MUTATION } from './gql';
+import { useCallback } from 'react';
+import { getStoredAccountInfo, handleGQLError } from '../utils';
+import { SigninLocationState } from '../interfaces';
+import { useFinishOAuthFlowHandler } from '../../../lib/oauth/hooks';
+import { useValidatedQueryParams } from '../../../lib/hooks/useValidate';
+import { SigninQueryParams } from '../../../models/pages/signin';
+import { ConsumeRecoveryCodeResponse, SubmitRecoveryCode } from './interfaces';
+import AppLayout from '../../../components/AppLayout';
+import CardHeader from '../../../components/CardHeader';
+
+export const viewName = 'signin-recovery-code';
+
+export type SigninRecoveryCodeContainerProps = {
+  integration: Integration;
+  serviceName: MozServices;
+};
+
+export const SigninRecoveryCodeContainer = ({
+  integration,
+  serviceName,
+}: SigninRecoveryCodeContainerProps & RouteComponentProps) => {
+  const authClient = useAuthClient();
+  const { finishOAuthFlowHandler, oAuthDataError } = useFinishOAuthFlowHandler(
+    authClient,
+    integration
+  );
+  // TODO: FXA-9177, likely use Apollo cache here instead of location state
+  const location = useLocation() as ReturnType<typeof useLocation> & {
+    state: SigninLocationState;
+  };
+  const { queryParamModel } = useValidatedQueryParams(SigninQueryParams);
+  const { redirectTo } = queryParamModel;
+
+  const signinLocationState =
+    location.state && Object.keys(location.state).length > 0
+      ? location.state
+      : getStoredAccountInfo();
+
+  const { sessionToken } = signinLocationState;
+
+  const [consumeRecoveryCode] = useMutation<ConsumeRecoveryCodeResponse>(
+    CONSUME_RECOVERY_CODE_MUTATION
+  );
+
+  const submitRecoveryCode: SubmitRecoveryCode = useCallback(
+    async (recoveryCode: string) => {
+      try {
+        // this mutation returns the number of remaining codes,
+        // but we're not currently using that value client-side
+        // may want to see if we need it for /settings (display number of remaining backup codes?)
+        const { data } = await consumeRecoveryCode({
+          variables: { input: { code: recoveryCode } },
+        });
+
+        return { data };
+      } catch (error) {
+        return handleGQLError(error);
+      }
+    },
+    [consumeRecoveryCode]
+  );
+
+  // TODO: UX for this, FXA-8106
+  if (oAuthDataError) {
+    return (
+      <AppLayout>
+        <CardHeader
+          headingText="Unexpected error"
+          headingTextFtlId="auth-error-999"
+        />
+      </AppLayout>
+    );
+  }
+
+  // TODO if previous page was force_auth, redirect there - else to signin
```

Comment:
Do we need a followup for this?

Reply:
Based on https://github.com/mozilla/fxa/pull/16387, I think I can remove this comment that was ported over from content-server. Looks like we support the force_auth route but it just directs to signin anyway.